### PR TITLE
Reject non-nullable aggregate types when deserializing method parameters

### LINKED ISSUE #5805: Getting Error for eth get transaction receipt
While doing a curl on the server for eth get transaction receipt 
`curl --data '{"method":"eth_getTransactionReceipt","params":[["0x80757153e93d1b475e203406727b62a501187f63e23b8fa999279e219ee3be71"]],"id":1,"jsonrpc":"2.0"}' -H "Content-Type: application/json" -X POST localhost:8545`

getting error as of below.
`{"jsonrpc":"2.0","error":{"code":-32603,"message":"Internal error","data":"System.NullReferenceException: Object reference not set to an instance of an object.\n   at Nethermind.Db.DbExtensions.Get(IDb db, Keccak key)\n   at Nethermind.Blockchain.Receipts.PersistentReceiptStorage.FindBlockHash(Keccak txHash)\n   at Nethermind.Blockchain.Receipts.FullInfoReceiptFinder.FindBlockHash(Keccak txHash)\n   at Nethermind.Facade.BlockchainBridge.GetReceiptAndEffectiveGasPrice(Keccak txHash)\n   at Nethermind.JsonRpc.Modules.Eth.EthRpcModule.eth_getTransactionReceipt(Keccak txHash)\n   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)\n   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)"},"id":1}`

### PR DESCRIPTION
**Pull request by emlautarom1**
Fixes #5805

## Changes

- Fail with `InvalidParams` when deserializing non-nullable aggregate types (ex. Arrays and Objects).

#### What types of changes does your code introduce?

- [x] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

#### Notes on testing

Tested the provided example in the referenced ticket.

#### Requires documentation update

- [ ] Yes
- [x] No

#### Requires explanation in Release Notes

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/Nethermind/Nethermind.JsonRpc/JsonRpcService.cs [hunk 1]
```diff
@@ -325,6 +325,10 @@ private void LogRequest(string methodName, string?[] providedParameters, Paramet
                     if (providedParameter.StartsWith('[') || providedParameter.StartsWith('{'))
                     {
                         executionParam = _serializer.Deserialize(new JsonTextReader(new StringReader(providedParameter)), paramType);
+                        if (executionParam is null && !IsNullableParameter(expectedParameter))
+                        {
+                            throw new JsonReaderException($"Cannot deserialize parameter `{providedParameter}` into type {paramType}");
+                        }
```
Comment (emlautarom1) [inline src/Nethermind/Nethermind.JsonRpc/JsonRpcService.cs:331]: Main change: if we try to deserialize an aggregate type like an array or an object and we get `null`, fail if the parameter is not nullable.

Unfortunately, our deserializer does not to provide enough information to differentiate deserializing `null` from failing to deserialize.

File: src/Nethermind/Nethermind.JsonRpc/JsonRpcService.cs [hunk 2]
```diff
@@ -325,6 +325,10 @@ private void LogRequest(string methodName, string?[] providedParameters, Paramet
                     if (providedParameter.StartsWith('[') || providedParameter.StartsWith('{'))
                     {
                         executionParam = _serializer.Deserialize(new JsonTextReader(new StringReader(providedParameter)), paramType);
+                        if (executionParam is null && !IsNullableParameter(expectedParameter))
+                        {
+                            throw new JsonReaderException($"Cannot deserialize parameter `{providedParameter}` into type {paramType}");
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/JsonRpcService.cs:330]: Shouldn't we just add `executionParameters.Add(Type.Missing)` ?

Comment (emlautarom1) [reply src/Nethermind/Nethermind.JsonRpc/JsonRpcService.cs:330]: I'm not sure, is an invalid parameter actually missing? From the users perspective, if we use the current approach, the result looks like:
```
{"jsonrpc":"2.0","error":{"code":-32602,"message":"Invalid params"},"id":1}
```

If we use `executionParameters.Add(Type.Missing)`, then the result looks like:
```
{"jsonrpc":"2.0","error":{"code":-32602,"message":"Parameter count mismatch.","data":"System.Reflection.TargetParameterCountException: Parameter count mismatch.
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Nethermind.JsonRpc.JsonRpcService.ExecuteAsync(JsonRpcRequest request, String methodName, ValueTuple`2 method, JsonRpcContext context) in /home/emlautarom1/Development/Nethermind/nethermind/src/Nethermind/Nethermind.JsonRpc/JsonRpcService.cs:line 186"},"id":1}
```

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.JsonRpc/JsonRpcService.cs:330]: Looking at this code: https://github.com/NethermindEth/nethermind/pull/6104/files/f5792fc8839aeea94d816b2266018a17591d2834#diff-bd27ffa6f72471764da77fa1d9e723ffb5eb275805fa6ee91380ee247d9d6549R295-R305

It should be `missing`. Leaking internal exception details here is also not great.

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.JsonRpc/JsonRpcService.cs]: It has its pro's and con's. I think it is fine for easier debugging user issues (For example, NullReferenceException on eth_createAccessList). But we shouldn't use exceptions on expected API errors.

**Review by LukaszRozmej — APPROVED** [review]

**Pull request merged**